exercises/sat/descr.md

```diff
@@ -0,0 +1,616 @@
+such an algorithm is known as a *SAT solver*.
+
+### CNF Conversion: Principles
+
+An arbitrary formula `f` can be converted to a CNF formula `f'` such that `f`
+and `f'` are *equisatisfiable*, that is, `f` is satisfiable if and only if
+`f'` is satisfiable. (We implement this conversion in the following question.)
+Thus, a SAT solver can be applied also to a formula that is not in conjunctive
+normal form.
+
+*Note.* One can obviously determine in linear time whether a formula in
+*disjunctive normal form* (DNF) is satisfiable. (Think about it.) However,
+converting an arbitrary formula to an equisatisfiable disjunctive normal form
+has exponential complexity, so the existence of such a linear-time algorithm
+is not of much interest.
+
+How does the CNF conversion work?
+
+First, note that it is easy to convert a formula `f` to an equivalent formula
+`f'` in negation normal form (NNF). The idea is to push every negation down
+towards the leaves. This exploits
+[De Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws):
+
+* `~(f1 /\ f2)` is equivalent to `~f1 \/ ~f2`.
+
+* `~(f1 \/ f2)` is equivalent to `~f1 /\ ~f2`.
+
+This transformation can be performed in linear time. In the following
+question, we do not implement it as a separate transformation; instead,
+we perform this transformation on the fly, at the same time as the CNF
+conversion.
+
+Second, to perform CNF conversion, there remains to eliminate the situations
+where a conjunction appears inside a disjunction, as in the formula `a \/ (b
+/\ c)`. The idea is to introduce a fresh *auxiliary variable* to stand for the
+problematic subformula: here, we introduce a new variable `x` to stand for the
+subformula `b /\ c`. In other words, we exploit the following fact:
+
+* `a \/ (b /\ c)` is equivalent to `exists x.( (a \/ x)  /\  (x <-> b /\ c) )`.
+
+In the main formula, `b /\ c` is replaced with `x`, so
+`a \/ (b /\ c)` becomes `a \/ x`, which is in CNF form,
+as desired. On the side, we use a double implication
+`x <-> b /\ c` to express the fact that `x` must have
+the same truth value as the conjunction `b /\ c`.
+
+A couple more remarks must be made for the above idea to really work.
+
+First, the use of a double implication `x <-> f` is problematic,
+because it represents the conjunction `(x -> f) /\ (f -> x)`,
+or `(~x \/ f) /\ (~f \/ x)`, where the subformula `f` is
+duplicated. Unless we are careful, this duplication could cause an
+exponential blowup in the size of formulae.
+Fortunately, there is in reality no need for a double implication:
+a single implication `x -> f` suffices. Indeed,
+
+* `a \/ (b /\ c)` is equivalent to `exists x.( (a \/ x)  /\  (x  -> b /\ c) )`.
+
+Please take some time to understand why this is true.
+
+Second, the existential quantification `exists x.(...)` is not really
+indispensable. Indeed, the above equivalence implies that:
+
+* the original formula `a \/ (b /\ c)` is satisfiable if and only if the
```

Comment:
Don't you need some freshness condition about `x`?

Reply:
Yes, probably `x` needs to be fresh. But I don't think it would help to say so; it would only complicate  the text, wouldn't it?

Reply:
Yes, probably.